wg-api/spec-documents/message-ports.md

```diff
@@ -0,0 +1,207 @@
+
+The [Channel Messaging API][channel-messaging-api] is a web standard that builds on `postMessage` by allowing the creation of multiple independent channels across which messages can be sent. On the web, it can be used to communicate with Web Workers, Service Workers, or between windows (or even within the same window, somewhat redundantly).
+
+Here's a short example of usage on the Web:
+
+```javascript=
+const {port1, port2} = new MessageChannel();
+port1.onmessage = (e) => console.log('port1 got message', e.data);
+someOtherWindow.postMessage(
+    `here's the other end of the channel`, '*', [port2]);
+
+// In the other window
+window.onmessage = (e) => {
+  const port2 = e.ports[0]
+  port2.postMessage('send a message over the channel')
+  // port1.onmessage will be triggered
+}
+```
+
+Under the hood, this is a light wrapper around Mojo pipes and the same algorithms that `window.postMessage` uses.
+
+### Node.js worker threads
+
+Node.js took inspiration from the Channel Messaging API when they designed the [communication system for worker threads](https://nodejs.org/dist/latest-v12.x/docs/api/worker_threads.html#worker_threads_class_messagechannel).
+
+### Transferables
+
+In the `postMessage` API, one of the options available when sending a message is to pass a list of _transferables_ to be passed along with the message. These objects, once transferred through `postMessage`, are no longer accessible to the sender, and ownership is transferred to the receiver. Some examples of transferable objects are `ArrayBuffer`, `ImageBitmap` and `MessagePort`.
+
+References to transferred objects can appear in the message data, and when deserialized, the message on the receiving end will contain references to those transferred objects in the appropriate place.
+
+## Proposal
+
+I propose that we add a new IPC primitive to Electron based on and interoperable with the Channel Messaging API.
+
+### New API surface
+
+#### ipcRenderer
+
+##### `ipcRenderer.postMessage(channel, message[, options])`
+
+* `channel` String
+* `message` any
+* `options` Object (optional)
+  * `transfer` any[] - a list of transferable objects to be sent with the message.
+
+This behaves similarly to `window.postMessage`, but instead of sending a message to another window, it sends the message to the main process.
+
+Also, this behaves similarly to `ipcRenderer.send`, but only takes a single argument after the channel (the message to be sent), and allows for additional options to be specified. In particular, it allows specifying a list of transferable objects to be sent along with the message.
+
+When this message is received in the main process, any `MessagePort` objects transferred will be deserialized as `ipcMain.MessagePort`.
+
+#### Class: WebContents
+
+##### `WebContents.postMessage(channel, message[, options])`
+
+* `channel` String
+* `message` any
+* `options` Object (optional)
+  * `transfer` any[] - a list of transferable objects to be sent with the message.
```

Comment:
There's a mismatch between this spec and the current PR https://github.com/electron/electron/pull/22532 which just supports `[list, of, transferables]`. I like the spec version better for a couple reasons:

1) More robust against adding more options in the future
2) More self documenting to explicitly say `{ transfer: [a, b, c] }`. I think this will especially help reduce confusion when we support passing ports in the `message`.

Looking at the history of the [web spec](https://html.spec.whatwg.org/multipage/web-messaging.html#posting-messages), it looks like the `postMessage(message, [transferables])` API was added first, and then the `postMessage(message, { transfer })` API was added later (they introduced a `targetOrigin` property that probably made them realize positional arguments can be painful for back-compat).

Reply:
I'm also keen to support the object as well as the array. It's listed in the PR as something to be done in a followup change.